### DISCUSSION TITLE
root access on clients

### DISCUSSION THREAD

**Discussion by Mogaba** [ltsp/ltsp#286] [Q&A]
How do you get root privileges on clients (chroot image)? It seems LTSP does something with /etc/shadow file so you can't login with local root account.

**Comment by alkisg** [comment]
Read the following sections from the [ltsp.conf man page](https://ltsp.org/man/ltsp.conf/):

* PWMERGE_SUR=, PWMERGE_SGR=, PWMERGE_DGR=, PWMERGE_DUR=: these explain how the user account merging is performed.
* POST_INIT_SET_ROOT_HASH: this, in the examples section, shows one way to set a root password in ltsp.conf instead of in the image. That way, the root password will work for all images and you don't have to set one in the chroot or in the server.

**Reply by bz-gt** [reply]
Hi, I am using Ubuntu 18.04 with LTSP and I have problem with applying POST_INIT_SET_ROOT_HASH. Edited ltsp.conf according to the example, but I am getting an error after calling ltsp initrd. The message says: 
```
: not foundtsp: 11: /usr/sbin/ltsp:
```
Image is generated, but system does not boot. My ltsp.conf:
```
[server]
MENU_TIMEOUT=-1
ADD_IMAGE_EXCLUDES="/etc/ltsp/image.excludes"
ADD_IMAGE_EXCLUDES="/usr/share/ltsp/server/image/image.excludes"

[clients]
#CRONTAB_01="0 19 * * * root /sbin/shutdown -h now"
POST_INIT_SET_ROOT_HASH="section_set_root_hash"


[set_root_hash]
sed 's|^root:[^:]*:|root:MY_ROOT_HASH_HERE:|' -i /etc/shadow

```
How can I resolve this?

**Comment by Mogaba** [comment]
I think I'm doing something wrong. Do I need to use PWMERGE_DUR for chroot accounts to appear in the client? Because this doesn't work:

[clients]
PWMERGE_DUR="user1"

I added user1 to the chroot image but there is no user1 in the shadow file in the client.

**Comment by alkisg** [comment]
First, did you run `ltsp initrd` to have the ltsp.conf change transferred to ltsp.img?
Second, are you sure you want "user1" to be a local user, unable to login via SSH?
Third, why didn't the POST_INIT_SET_ROOT_HASH work for your needs? What exactly are you trying to do, that users in ltsp.conf do not cover it?

**Comment by Mogaba** [comment]
Yes, I did `ltsp initrd`.
I'm just trying to understand how to add local users (and how to enable local root).
POST_INIT_SET_ROOT_HASH works, yes, but it seems a bit weird - editing shadow file manually. Is this the right way to add local users? And what exactly does PWMERGE_DUR do?

**Comment by alkisg** [comment]
Where does /home/username come from, for your "local user"? SSHFS, NFS, a local disk partition, or do you plan to include that in the squashfs image?

The normal place to put users in ltsp is in ltsp.img, not in the chroot; that way it will work with all images.

**Comment by Mogaba** [comment]
> The normal place to put users in ltsp is in ltsp.img, not in the chroot; that way it will work with all images

Sorry, I'm a bit confused, did I get this right: "ltsp image" removes chroot users (unless --cleanup=0), "ltsp initrd" makes ltsp.img that contains server's users, then users from both images merge together during boot?

**Comment by alkisg** [comment]
That is correct.
Note that there are no shadow entries in ltsp.img, so this isn't a security issue.

**Comment by alkisg** [comment]
@bz-gt, can you copy/paste the exact terminal output from the `sudo ltsp initrd` command?

I guess you have something wrong in ltsp.conf, which doesn't show up in your paste. For example, maybe by mistake you put a single quote inside the root hash, like `sed 's|^root:[^:]*:|root:wrong-quote-here'-hash-continues:|' -i /etc/shadow`. Or maybe you forgot to close some quotes earlier in ltsp.conf.

**Comment by bz-gt** [comment]
I pasted here whole ltsp.conf content. POST_INIT_SET_ROOT_HASH things were copy-pasted from documentation, I only replaced hash, which I copied from my /etc/shadow. I checked hash now and it contains only letters, digits, forward slashes, dollar signs and periods. Exact terminal output:
```
root@company-HVM-domU:~# ltsp initrd
: not foundtsp: 11: /usr/sbin/ltsp:
316 bloków
Generated ltsp.img:
-rw-r--r-- 1 root root 161792 lut 10 13:03 /srv/tftp/ltsp/ltsp.img
```

**Comment by alkisg** [comment]
That `not foundtsp` is clearly a typo.
Does ltsp initrd work properly if you remove ltsp.conf?
Does it work if you remove half of its contents?
Try to find the line that breaks it.

**Comment by bz-gt** [comment]
I noticed it but don't know where it comes from. Sure, when I remove everything below commented out 'CRONTAB' line, initrd is not raising any errors and OS is booting properly on client terminals.

**Reply by bz-gt** [reply]
@alkisg OK, I finally got it working. Honestly, I don't know how it happened, but I tried to create this file again from scratch and it just works. It looks the same, regret I didn't make diff. Only thing that changed is editor, so I wonder if it had something to do with character encoding. Anyways, thank you for your assistance, you were right that one of the lines is breaking this.
